File: backend/internal/protocol/bounds.go
package protocol

import (
	"encoding/hex"
	"strings"

	"github.com/hyunu/sentinel/internal/models"
)

func fieldUsesRemaining(field models.FieldSpec) bool {
	if field.LengthMode == "remaining" {
		return true
	}
	if field.Repeat == "until_end" {
		return true
	}
	return field.Type == "raw" && field.Length == 0 && !isBitField(field)
}

func readRemainingField(data []byte, cur *parseCursor, field models.FieldSpec) (interface{}, error) {
	cur.alignByte()
	end := cur.boundEnd(len(data))
	if cur.byteOff >= end {
		cur.byteOff = end
		cur.bitOff = 0
		if field.Type == "ascii" {
			return "", nil
		}
		return "", nil
	}
	slice := data[cur.byteOff:end]
	cur.byteOff = end
	cur.bitOff = 0

	switch field.Type {
	case "ascii":
		return strings.TrimRight(string(slice), "\x00"), nil
	case "hex", "raw", "dynamic":
		return hex.EncodeToString(slice), nil
	default:
		if len(slice) == 0 {
			return nil, nil
		}
		spec := field
		spec.Length = len(slice)
		spec.LengthMode = ""
		val, _, err := readField(slice, 0, spec)
		return val, err
	}
}

func newBoundedCursor(dataLen int) *parseCursor {
	return &parseCursor{maxByteOff: dataLen}
}

File: dashboard/src/lib/bounds.ts
import type { FieldSpec } from '../api';
import { alignByte, boundEnd, isBitField, type ParseCursor } from './bits';
import { bytesToHex, readField } from './fieldRead';

export function fieldUsesRemaining(field: FieldSpec): boolean {
  if (field.length_mode === 'remaining') return true;
  if (field.repeat === 'until_end') return true;
  return field.type === 'raw' && (!field.length || field.length === 0) && !isBitField(field);
}

export function readRemainingField(data: number[], cur: ParseCursor, field: FieldSpec): unknown {
  alignByte(cur);
  const end = boundEnd(cur, data.length);
  if (cur.byteOff >= end) {
    cur.byteOff = end;
    cur.bitOff = 0;
    return field.type === 'ascii' ? '' : '';
  }
  const slice = data.slice(cur.byteOff, end);
  cur.byteOff = end;
  cur.bitOff = 0;

  switch (field.type) {
    case 'ascii':
      return String.fromCharCode(...slice).replace(/\x00+$/, '');
    case 'hex':
    case 'raw':
    case 'dynamic':
      return bytesToHex(slice);
    default:
      if (slice.length === 0) return null;
      return readField(slice, 0, { ...field, length: slice.length, length_mode: undefined }).value;
  }
}
